### ISSUE TITLE
Maven fails to build after ozone ufs upgrade to 1.0.0

### ISSUE THREAD

**Issue by Xenorith** [Alluxio/alluxio#12338]
**Alluxio Version:**
2.4.0

**Describe the bug**
After merging the PR to bump ozone ufs version (https://github.com/Alluxio/alluxio/pull/12226), the following maven command hangs:
`mvn -T 4C clean install -Dmaven.javadoc.skip=true -DskipTests -Dlicense.skip=true -Dcheckstyle.skip=true -Dfindbugs.skip=true -P'\''!ufs-hadoop-1.2,!ufs-hadoop-2.2,!ufs-hadoop-2.6'\''`

When I run jstack on the process, the one running thread approximately has the following stacktrace:
```
"BuilderThread 17" #60 prio=5 os_prio=0 tid=0x00007fcea0106800 nid=0xbaa runnable [0x00007fcdb9c2a000]
   java.lang.Thread.State: RUNNABLE
        at org.jdom2.Element.isAncestor(Element.java:1052)
        at org.jdom2.ContentList.checkPreConditions(ContentList.java:222)
        at org.jdom2.ContentList.add(ContentList.java:244)
        at org.jdom2.Element.addContent(Element.java:950)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.insertAtPreferredLocation(MavenJDOMWriter.java:292)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.iterateExclusion(MavenJDOMWriter.java:488)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.updateDependency(MavenJDOMWriter.java:1335)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.iterateDependency(MavenJDOMWriter.java:386)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.updateModel(MavenJDOMWriter.java:1623)
        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.write(MavenJDOMWriter.java:2156)
        at org.apache.maven.plugins.shade.pom.PomWriter.write(PomWriter.java:75)
        at org.apache.maven.plugins.shade.mojo.ShadeMojo.rewriteDependencyReducedPomIfWeHaveReduction(ShadeMojo.java:1049)
        at org.apache.maven.plugins.shade.mojo.ShadeMojo.createDependencyReducedPom(ShadeMojo.java:978)
        at org.apache.maven.plugins.shade.mojo.ShadeMojo.execute(ShadeMojo.java:538)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:200)
        at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:196)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

This seems to be an issue with the maven shading plugin. Looking up `        at org.apache.maven.plugins.shade.pom.MavenJDOMWriter.insertAtPreferredLocation(MavenJDOMWriter.java:292)`, I found that other people have encountered a similar issue:
- https://stackoverflow.com/questions/52221326/maven-shade-plugin-failed-to-execute-goal
  - but it seems like our plugin version is sufficiently up to date
- https://github.com/apache/spark/pull/23061
  - this seems to avoid the issue by removing parallelism, but this would drastically slow down our builds

**To Reproduce**
See above mvn command

**Expected behavior**
Maven command completes

**Urgency**
Offending PR has been reverted to avoid build issues (https://github.com/Alluxio/alluxio/pull/12337)

**Comment by maobaolong** [comment]
@Xenorith Thank you for report this issue, it seems is the same issue like this https://issues.apache.org/jira/browse/MSHADE-148.

There are 2 options to work around.
- Patch https://github.com/apache/maven-shade-plugin/pull/16/ and use the patched version maven-shade-plugi.
- Add `<createDependencyReducedPom>false</createDependencyReducedPom>` to the `alluxio/shaded/client/pom.xml` to avoid generated DependencyReducedPom.xml

**Comment by Xenorith** [comment]
Thanks for looking into this so quickly @maobaolong 

I am not familiar enough with maven to help decide which option is best to work around the issue. I would strongly prefer we avoid `Removing parallelism` because it would drastically slow down build times. Otherwise, I cannot say much about the other options you proposed. 

@apc999 Can you provide some guidance on what is the best way to proceed with this issue?

**Comment by maobaolong** [comment]
@Xenorith Sure, we need to support parallelism build, there are one more option

The SPARK project is suffering from this issue too, https://issues.apache.org/jira/browse/SPARK-8819 this is a option to support parallelism build with  let `createDependencyReducedPom=false` default, but support a `release flag` to let `createDependencyReducedPom=true` and non-parallelism build for release build, this is we don't need to create dependency deduced pom file every time, when we need, just with `release flag`.

**Comment by apc999** [comment]
@maobaolong @Xenorith  Thanks for getting to the bottom of this problem!
To me removing reduced pom (like https://stackoverflow.com/a/11322920/3161750) is a good way to go as it seems not really used at all (https://stackoverflow.com/a/35155294/3161750)

@calvinjia  any input?

**Comment by calvinjia** [comment]
@apc999 The dependency reduced pom is relevant when published so I think @maobaolong ’s proposed solution is the best of both worlds.

**Comment by apc999** [comment]
I actually think the generated pom in src dir is not relevant to depending modules. There will be a pom in `.m2/repository/org/alluxio/alluxio-shaded-ozone/2.4.0` which is identical to the reduced dependency pom. maybe we can test it to have `createDependencyReducedPom=fase` and see if `.m2/repository/org/alluxio/alluxio-shaded-ozone/2.4.0/alluxio-shaded-ozone-2.4.0.pom` is identical to initial pom or the generated pom @calvinjia

**Comment by calvinjia** [comment]
Sure, my understanding is with false it will have the original pom

**Comment by maobaolong** [comment]
@apc999 I have checked that if we set `createDependencyReducedPom=false`, the `.m2/repository/org/alluxio/alluxio-shaded-ozone/2.4.0/alluxio-shaded-ozone-2.4.0.pom` will be the same as `alluxio/shaded/ozone/pom.xml`

**Comment by apc999** [comment]
This issue has been fixed by  #12367
